src/prefect/server/utilities/messaging/memory.py

```diff
@@ -29,23 +30,35 @@
 class MemoryMessage:
     data: Union[bytes, str]
     attributes: Dict[str, Any]
+    retry_count: int = 0
 
 
 class Subscription:
     topic: "Topic"
     _queue: asyncio.Queue
     _retry: asyncio.Queue
 
-    def __init__(self, topic: "Topic") -> None:
+    def __init__(self, topic: "Topic", max_retries: int = 3) -> None:
         self.topic = topic
+        self.max_retries = max_retries
+        self.dead_letter_queue = list()
         self._queue = asyncio.Queue()
         self._retry = asyncio.Queue()
 
     async def deliver(self, message: MemoryMessage) -> None:
         await self._queue.put(message)
 
     async def retry(self, message: MemoryMessage) -> None:
-        await self._retry.put(message)
+        message.retry_count += 1
+        if message.retry_count > self.max_retries:
+            logger.warning(
+                "Message failed after %d retries and will be moved to the dead letter queue",
+                message.retry_count,
+                extra={"event_message": message},
+            )
+            self.dead_letter_queue.append(message)
```

Comment:
The DLQ here could end up growing without bounds.  What do you think about writing these to a file in `$PREFECT_HOME` instead?  That gives the user a chance to review them and decide whether there's something that needs to happen, or ignore them and drop them.  This also gives us an opening to implement some kind of redrive process for them later.

Either way, we probably don't want to store them in-memory indefinitely, so we need a way to purge these if we don't write them to disk.

Reply:
That's a good idea! For other message brokers, I think we might want to have the DLQ in the same infra (e.g. Redis), but that doesn't work as well for the in-memory implementation. I'll update the current implementation to write to disk with an eye toward writing other places in the future.

Reply:
yep exactly ^